src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -278,7 +281,7 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, Diagnosti
                         sourceTuple.Syntax,
                         sourceTuple,
                         wasTargetTyped: false,
-                        sourceTuple.Arguments.SelectAsArray(e => BindToNaturalType(e, diagnostics, reportDefaultMissingType)),
+                        sourceTuple.Arguments.SelectAsArray((e, arg) => arg.self.BindToNaturalType(e, arg.diagnostics, arg.reportDefaultMissingType), (self: this, diagnostics, reportDefaultMissingType)),
```

Comment:
SelectAsArray is no longer worth it, IMO. Time to put into a foreach.

Reply:
This call is not the performance problem. The performance problem is caused by this lambda expression forcing the arguments to the method to be captured _even when this line will not be hit_.

Reply:
Right, not a performance problem, this is just bad code. It's unclear what's going on and it would be clearer as a foreach.

Reply:
```C#
var boundArguments = ArrayBuilder<BoundExpression>.GetInstance(sourceTuple.Arguments.Length);
foreach (var arg in sourceTuple.Arguments)
{
    boundArguments.Add(BindToNaturalType(arg, diagnostics, reportDefaultMissingType);
}
...
boundArguments.ToImmutableAndFree(),
```

Reply:
Rebased to eliminate this change. Will file a follow-up issue.

Reply:
➡️ #41054